lib/main.dart

```diff
@@ -7,14 +7,14 @@ import 'model/binding.dart';
 import 'notifications/receive.dart';
 import 'widgets/app.dart';
 
-void main() {
+Future<void> main() async {
   assert(() {
     debugLogEnabled = true;
     return true;
   }());
   LicenseRegistry.addLicense(additionalLicenses);
-  LiveZulipBinding.ensureInitialized();
   WidgetsFlutterBinding.ensureInitialized();
+  await LiveZulipBinding.ensureInitialized();
```

Comment:
The goal for this change is to pre-fetch the `deviceInfo` and `packageInfo` while ZulipBinding is being initialized and then later they can be used from [non-async functions](https://github.com/zulip/zulip-flutter/blob/8458b725ae0a9f6c7fe1cfea2ec71b4ae183b3ff/lib/widgets/content.dart#L1208-L1244) via the getters.

Reply:
That make sense from an ergonomics perspective; but it makes me nervous to be introducing this wait into the startup path. Probably on any reasonable system the queries will return quickly and it's fine, but
* there may be bugs in these third-party plugins that cause them to sometimes hang before supplying an answer;
* there may be devices out there where some bug in a vendor's low-quality modifications to the Android platform software cause the queries to hang, or just to be slow, even though the plugins are doing only perfectly reasonable things that totally ought to be reliably fast.

Then the thing about putting awaits like these into the startup path is that it'd escalate any bugs of those types into a serious problem in the Zulip app for affected users.

So let's arrange things instead so that if for some reason those queries are slow, or even hang forever and never return, everything fails gracefully — just the places that actually wanted to use the answers have to make do with not knowing the answers.

One good way to do that, API-wise, would be to indeed make those into nice non-async getters, but returning nullable. Then when the binding is initialized, it kicks off the queries, and if the queries have finished by the time we try to make a request then we use the results; if not, we fall back to not knowing this information.

Reply:
I guess most of the time we'll be firing off one request promptly after startup, the register-events request. So if we find that that often happens before the platform queries finish, then we can add an extra wrinkle where we wait for the queries to finish… but with a timeout, like 10ms or something, so we prioritize the user's experience over being sure we include this information.

Let's keep that as a separate followup commit, though. Or it'd be great if you try a little manual experimenting to see if that's a problem in practice; if it seems like it isn't, we can skip that wrinkle entirely, and keep it in our back pocket as a possible followup for later if it turns out to be a problem sometimes in wider deployment.

Reply:
Thanks, yeah we definitely should have a fail safe if plugins fails for some reason or are slow to retrieve the data.

The latest revision now doesn't `await` on pre-fetchs for both (`deviceInfo` and `packageInfo`), and where the getters are called there's now some kind of relevant fallback just-in-case.

> if you try a little manual experimenting to see if that's a problem in practice

I tried that, by [`throw`-ing in case either of infos are null](https://github.com/zulip/zulip-flutter/blob/a9cf8d3527e5a27b2fe983f6709176620edad90b/lib/model/binding.dart#L353) but didn't see any exceptions in both release and debug builds for Android & macOS (can't try release on iOS Simulator) during startup.